### DISCUSSION TITLE
[Complete] RFC: Route configuration for hybrid rendering

### DISCUSSION THREAD

**Discussion by alan-agius4** [angular/angular#56785] [RFCs]
**Author:** @alan-agius4
**Contributors:** @AndrewKushnir, @dgp1130 
**Posted:** 1st July 2024
**Status:** Closed
# Introduction

As Angular continues to advance, it is crucial to enhance its hybrid rendering capabilities. These enhancements address the growing complexity and performance demands of modern web applications. Server-side rendering and build-time prerendering provide significant benefits, such as improved performance, better SEO, and a smoother user experience by delivering fully rendered pages quickly.

To fully leverage these capabilities, there is a need to provide additional route information. This includes page rendering mode, additional response headers, response status codes, and more. These enhancements enable more fine-grained control over how each route is handled and rendered, allowing developers to optimize their applications for different scenarios and requirements.

By introducing these new capabilities, we aim to resolve several issues reported by the community, including the need for better dependency injection (DI) token handling, asynchronous route declarations for static prerendering, and reduced dependencies on Node.js. Additionally, these improvements will address challenges related to integrating SSR with internationalization (i18n), supporting SSR entry points in Angular's build tools, and more.

The naming used in this document is not final and is intended to accurately describe the features being discussed. The final API design and naming will be determined in a later phase taking into consideration the input from this RFC.

# Overview

Currently, we automatically render server-side pages with routing parameters and prerender non-parameterized pages. We offer configuration to override this behavior and this proposal intends to make it more granular and easier to configure, while also enabling prerendering (SSG) for parameterized routes.

Initially, our aim is to provide developers with the following configuration options per route:
- Page rendering mode
  - **SSR:** The page renders on the server during a request
  - **SSG:** The page renders during build time and is served as a static asset
  - **CSR:** The page renders in the browser
- SSG handler for routes with dynamic parameters

Looking forward, we plan to further extend this to include:
- Additional response headers
- Response status code such as redirects
- Incremental Static Regeneration (ISR) and Stale-While-Revalidate (SWR) rendering modes

Improvements proposed in this RFC address a number of reported issues:
- [Providing a DI token in `commonEngine.render()` does not provide its value to application](https://github.com/angular/angular-cli/issues/26323)
- [Enable declaration of async routes for static prerendering](https://github.com/angular/angular-cli/issues/25721)
- [Exclude routes for pre-rendering](https://github.com/angular/angular-cli/issues/25728)
- [Reduce dependencies on nodejs for the `CommonEngine`](https://github.com/angular/angular-cli/issues/25718)
- [SSR with i18n not working](https://github.com/angular/angular-cli/issues/25726)
- [Support use ssr entrypoint in `@angular-devkit/build-angular:dev-server`](https://github.com/angular/angular-cli/issues/27144)

### Challenges

1. **Integrating Server-Specific Code:** To extend the Angular router and existing `Route` interface with additional options, server-specific code must be integrated into the client-side codebase. This introduces complexity that would rely on techniques to eliminate server-only code from client bundles.
1. **Code Elimination Techniques:** Relying on code elimination techniques to remove server-specific logic from the route configuration, especially when you need to invoke server-only code raises security concerns about accidentally retaining secrets in the client-side bundle.
1. **Configuration Management:** Managing nested route configurations to determine whether options set in the parent route should apply to child routes adds another layer of complexity.

### Proposed solution

To address these challenges, we propose implementing a separate route configuration specifically tailored for the server. There is no need to re-declare your routes, you can only add ones that require some specific server behavior.

This distinct configuration is kept separate from the Angular router, ensuring a clear separation between server-side and client-side code. This separation helps maintain clarity and simplifies the management of routing concerns.

```typescript
type RenderMode = 'prerender' | 'server' | 'client';

interface ServerRouteConfig {
  path: string;
  mode: RenderMode;
  getPrerenderPaths?():
    | Promise<Record<string, string>[]>
    | Record<string, string>[];
}
```

### Example configuration

**Angular application route configuration (unchanged)**

```typescript
import { Routes } from '@angular/router';

export const routes: Routes = [
  { path: '', ... },
  { path: 'login', ... },
  { path: 'contact', ... },
  { path: 'about-us', ... },
  { path: 'policy', ... },
  { path: 'product/:category/:id', ... },
];
```

**Server route configuration**

```typescript
import { ServerRouteConfig } from '@angular/ssr';

export const serverRouteConfig: ServerRouteConfig[] = [
  { path: '/login', mode: 'server' },
  { path: '/**', mode: 'prerender' },
];
```

While not all routes need to be re-defined in the server configuration, those that are defined must match the routes provided in the Angular router. Discrepancies between the server configuration and the Angular router will result in build-time errors.

In the above example, all routes will be prerendered except for the `/login` route, which will be server-side rendered. The `/**` will act as a catch-all, so the server route configuration does not need to define all the routes explicitly.

# What kind of feedback are we looking for?

While we're eager to hear any and all comments on the proposed changes, there are a few specific discussion points we are particularly interested in. Specifically, we would appreciate feedback on the following:

- **Discussion Question 1: Build-Time Restrictions**, do any of the specified build-time restrictions for i18n applications present significant obstacles for your scenarios?
- **Discussion Question 2: Integration of Server Routing Config with Application Routing**, how effectively does the server routing config file integrate with the application routing file?
- **Discussion Question 3: Missing Use Cases**, are there any use cases where this proposed solution might not work or could present challenges?
- **Discussion Question 4: Response / Request objects**, when working with response and request objects, are there any use cases where normalizing them to the web version could present challenges?
- **Discussion Question 5: Localized URL Pathnames**, Do you presently localize URL pathnames? If yes, how do you manage them currently?

When you leave a comment, please specify the question(s) to which you are responding. As always, keep our code of conduct in mind. We understand that a proposal of this scope significantly impacts Angular and how you use it in your projects. We appreciate your commitment and investment in this project and ask that you keep comments respectful.

# Proposal

## Rendering mode

Initially, support will be provided for three rendering modes: SSR, SSG and CSR. This allows developers to have precise control over how each route is rendered, offering an easy way to exclude certain routes from prerendering. For more details, refer to the related [issue](https://github.com/angular/angular-cli/issues/25728).

Upon receiving a request, the server matches it with the route configuration. If applicable, the server will serve either the pre-rendered SSG file, an HTML document for CSR, or SSR the page directly.

```typescript
import { ServerRouteConfig } from '@angular/ssr';

export const serverRouteConfig: ServerRouteConfig[] = [
  { path: '/contact-us', mode: 'server' },
  { path: '/login', mode: 'client' },
  { path: '/**', mode: 'prerender' },
];
```

## SSG configuration for routes with dynamic parameters

Currently, to enable SSG for a route with dynamic parameters (e.g., `/product/:id`), [developers must create a separate `routes.txt` file in advance](https://angular.dev/guide/prerendering#prerendering-parameterized-routes) and populate it with a list of all possible routes.

The proposed API streamlines the process of prerendering parameterized routes, which typically require data fetching from external sources or databases, such as a Content Management System (CMS). With the `getPrerenderPaths` function, developers can define which paths need data retrieval and subsequent pre-rendering. This function is executed exclusively during the build phase and allows to provide parameterized values necessary for pre-rendering a route. The function is invoked in an injection context, so you can use Angular DI there via the \`inject\` function.

```typescript
import { ServerRouteConfig } from '@angular/ssr';

export const serverRouteConfig: ServerRouteConfig = [
  {
    path: '/product/:id',
    mode: 'prerender',
    async getPrerenderPaths() {
      const dataService = inject(ProductService);
      const ids = await productService.getIds(); // ['1', '2', '3']

      return ids.map(id => ({ id })); // `id` is used in place of `:id` in the route path.
    },
  },
];
```

In the above case, `/product/1`, `/product/2`, and `/product/3` will be prerendered.

## Request and Response access via DI tokens

We are introducing new public-facing APIs designed to streamline common scenarios, where access to the Request and Response instances is required during SSR. These APIs will be automatically configured when you run the `ng new --ssr` or `ng add @angular/ssr` commands.

When integrating Angular SSR, it is important to note several key differences and considerations:
- Node.js File System APIs may be unavailable or inaccessible in non-Node.js environments, requiring files to be read using provider-specific APIs, inlined as data, or transformed into JavaScript files.
- Some environments utilize the Web [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) and [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) APIs, while others use a variant version of Node.js [Request](https://nodejs.org/api/http.html#class-httpincomingmessage) and [Response](https://nodejs.org/api/http.html#class-httpserverresponse) APIs.

These new APIs ensure that Angular SSR can be effectively utilized across different environments, maintaining functionality and performance.

### Injection tokens

- `REQUEST`: This will be of type [`Request`](https://developer.mozilla.org/en-US/docs/Web/API/Request) and will include the request variant specified in the Web specifications.
- `RESPONSE`: This will be of type [`ResponseInit`](https://developer.mozilla.org/en-US/docs/Web/API/Response/Response#options) and will contain the options specified in the Web specifications for initializing a response.
- `REQUEST_CONTEXT`: This token is of type `unknown`, designed to allow third-party implementations to include additional information not present in the main request. For example, [Netlify's context](https://docs.netlify.com/edge-functions/api/#netlify-specific-context-object), which includes extra geo-location data. Using the `unknown` type in TypeScript is necessary to prevent dependence on a specific third-party implementation. This approach strikes a balance between flexibility and type safety.

```typescript
import { REQUEST, RESPONSE } from '@angular/ssr';

@Component({})
export class MyComponent {
  constructor() {
    const request = inject(REQUEST);
    const header = request?.headers.get('x-custom');

    const response = inject(RESPONSE);
    if (response) {
      response.status = 404;
    }
  }
}
```

## Build-time i18n

One key challenge users encounter involves integrating SSR with build-time i18n. This API aims to address this challenge by efficiently managing i18n. However, several important considerations arise. Each locale corresponds to a distinct application within its bundle, all rendering within a shared global context. This design requires the application to follow some constraints to work effectively.

### Restrictions

- `$localize` will not work in `server.ts`.
- `server.ts` executes outside an Angular application and injection context, therefore Angular DI tokens/providers are not accessible within that file.
- The Angular application exported from `main.server.ts` or other DI tokens must not be referenced in the `server.ts` instead, the Angular CLI will inject references to the localized application bundle.

# FAQs

**Q: Why use the Web spec Request and Response types?**
This approach prevents individual components and services from being dependent on the specific request/response types of your server infrastructure, which is a valuable benefit.

**Q: Instead of using an additional routing file, why not use file-based routing?**
File-based routing is something we want to explore in the future.

**Q: With this new API, will I be able to use my API defined in `server.ts` with `ng serve`?**
Yes. One of the goals of this API is to make this possible.

**Q: Will inlining of critical CSS be available for non-Node.js environments?**
Yes, this API will handle this use case as well.

**Q: Will `RESPONSE` and `REQUEST` tokens be available during prerendering?**
No, since there are no requests during prerendering, these tokens will be `undefined`.

**Comment by imaksp** [comment]
Can you provide more clarity about how to use server only env variables in `getPrerenderPaths` (inside ProductService).

also it would be good to add `getPrerenderData` or `getStaticProps` or `resolve` function support in server router config, it will be helpful in fetching component data using server only secret. (without using _ExperimentalPendingTasks_ inside of the component)

**Reply by alan-agius4** [reply]
> Can you provide more clarity about how to use server only env variables in `getPrerenderPaths` (inside ProductService).

In the above example, `ProductService` functions as a standard Angular service. Should you require an environment variable within the `getPrerenderPaths` method, you can access it using `process.env`. The, `getPrerenderPaths` runs during build time. Hence, to utilize an environment variable, you must pass it through `ng build` and retrieve it via `process.env`.

For example:

```bash
$ MY_VAR="value" ng build
```

```typescript
async getPrerenderPaths() {
    const envVar = process.env['MY_VAR'];
    // Use envVar as needed
    // ...
}
```

> also it would be good to add getPrerenderData or getStaticProps or resolve function support in server router config, it will be helpful in fetching component data using server only secret. (without using ExperimentalPendingTasks inside of the component)

While this can potentially be a consideration as a future enhancements, it is currently not included in the initial proposal at this stage.

**Reply by imaksp** [reply]
Ok, please consider it for future, it will also remove data fetching code from client bundle. & we can safely use server secrets for getting static or dynamic component data.  Thanks.

**Reply by alan-agius4** [reply]
I now understand your case better, and this capability is already possible today, so there's no necessity for additional hooks. Dependency injection enables the creation of server-only services, as discussed [here](https://github.com/angular/angular/issues/53168#issuecomment-1827805974).

**Reply by imaksp** [reply]
Hi, checked quickly, but I think it will still require checking `isPlatformServer` condition inside of component constructor & inside of condition we can call this service fn & after that we will need to add it in  `ExperimentalPendingTasks` , 
resolve support in server route will simplify it. we can just call this server service inside it & return data, Angular will wait for it to complete at build time if route mode is `prerender`.

**Reply by diesieben07** [reply]
From how I understand this proposal, the "servier side route configuration" _augments_ the existing route configuration. So you would just do your data fetching like normal in `resolve`.

**Reply by imaksp** [reply]
@diesieben07  which resolve ? if I do it in client route config it will run every time on client when route changes, & it will also expose server only secrets. I am talking about server only resolve function. & with prerender mode it should run only at build time, or when ISR is initiated.

**Reply by diesieben07** [reply]
Even with SSG enabled for the page, it can still be rendered on the client if a client-side navigation happens: user types in `/route-a` but then clicks on a `routerLink` on that page that goes to `route-b`. Even if `route-b` is configured for SSR or SSG it will still be rendered on the client. Your `resolve` functions need to handle this scenario.

**Reply by imaksp** [reply]
Hi, with SSG server will directly return cached HTML on first navigation even if navigation occurs after an initial render  & data won't be updated unless it is built again. this is how I understand it. & it was working this like this with Next.js pages dir. client config won't have any resolve function in this case.

**Reply by diesieben07** [reply]
That's not how it works in Angular. Once the client-side app is hydrated after initial navigation the server (or any pre-generated HTML) is not involved anymore.

**Reply by imaksp** [reply]
Ok, then it should change with this hybrid rendering support RFC. it could preload SSG page in advance to like in Next, automatically based on `routerLink` presence or manually by using some new function. & content should update only during build or with on-demand ISR call.

**Reply by anthony-b-dev** [reply]
I'm also in favor of Angular intelligently knowing that if a page has been prerendered, then navigation to that route should pull that statically generated content. Or there should at least be a flag in the route definition that leads to this behavior. 

Is there a reason why this is undoable or undesirable?

**Reply by alan-agius4** [reply]
> I'm also in favor of Angular intelligently knowing that if a page has been prerendered, then navigation to that route should pull that statically generated content. Or there should at least be a flag in the route definition that leads to this behavior.
> 
> Is there a reason why this is undoable or undesirable?

While this idea is interesting, this feature is outside the scope of this RFC.

**Reply by imaksp** [reply]
Hi, if pre-rendered HTML is not used after initial render, we will need to set data manually in `TransferState` after calling server only private API from root component & use it to support CSR case for such route (on client navigation). 
if data fetching is supported in server route config it would simplify the component code. I hope this gets included in the future.

**Comment by anthony-b-dev** [comment]
For SSG routes with dynamic parameters, it would be good to be able to support multiple dynamic parameters.

Ie `products/:id/details/:variantId`

Maybe not the cleanest of design practices, but we should still be able to do it, or put a limit or something on how many we can do.

**Reply by alan-agius4** [reply]
This will be supported.

**Comment by MarcoGlauser** [comment]
I really like the idea to move away from freestyle server.ts into a more structured way of handling common ssr challenges.

Our SSR rendering is quite slow (~1-3 sec) so we have a lot of caching in place. (SSR performance issues are difficult to track down)
Currently, we inject a redis cache provider in the `server.ts` but not in the client application config. Would that still be possible with the new proposal?

**Reply by alan-agius4** [reply]
Certainly, you're still able to customize your `server.ts` file and include additional middlewares as needed.

**Comment by manfredsteyer** [comment]
Thanks for this RFC. It looks exciting!

I understand why we need a separate server-side routing file, but I'd like to know if we can ensure somehow that it aligns with the client-side one, e.g., via the Angular Language Server.

I'm also wondering what the server-side counterpart of a lazy client-side route definition is. Will there be a server-side router configuration for each client-side route definition, or will we put everything into one server-side router configuration?

**Reply by alan-agius4** [reply]
As illustrated in the example above, not all app routes need to be defined in the server routes configuration, but all server configuration routes must match an Angular router route.

The server configuration adds additional metadata to the Angular Router routes. Therefore, the server route configuration does not require lazy-loading, as this is a concept specific to the Angular router.

**Comment by robertIsaac** [comment]
1. how would lazy loaded routes be configured in the server?
let's say I have `question` lazy loaded route with two nested routes 
a. `:id` I want it to be SSR
b. `:id/edit` that I want it to be CSR
2. what's the default mode for a route?
or is `  { path: '/**', mode: '...' }` required?

**Reply by alan-agius4** [reply]
1. There are a couple of ways to achieve this, depending on the default rendering mode of the application.

```typescript
export const serverRouteConfig: ServerRouteConfig[] = [
  { path: '/question/**/edit', mode: 'csr' },
  { path: '/**', mode: 'server' }
];
```

```typescript
export const serverRouteConfig: ServerRouteConfig[] = [
  { path: '/question/**/edit', mode: 'csr' },
  { path: '/question/**', mode: 'server' },
  { path: '/**', mode: 'csr' }
];
```

2. The `{ path: '/**', mode: '...' }` route is used as the default rendering mode. While we could enhance the current `prerender` and `ssr` application builder options in `angular.json`, offering multiple ways to achieve the same goal can sometimes lead to user confusion.

**Reply by robertIsaac** [reply]
you misunderstood my section question
I will ask it in a better way

if my server routing is only
```
export const serverRouteConfig: ServerRouteConfig[] = [
  { path: '/question/**/edit', mode: 'csr' },
];
```
what mode will be `/question/**`?

**Reply by alan-agius4** [reply]
That will cause a build-time error because none of the server routes correspond to the routes defined in Angular.

**Comment by KhaledGabr** [comment]
Thanks for the RFC, indeed most applications will be using a combination of these rendering strategies based on the route. 

I have 3 points:

1) Local development:
Currently I have two "ng serve" commands one with SSR and one without. Because I had some SSR routes and some CSR, which makes development quite hard, as I kept switching between the two served application during development. Would that no longer be needed, and we will have a single serve commands for both. 

2) Build and Deployment
During build though, I have two build commands, one for SSR (which deploys to a node server) and one for CSR (which deploys to a S3 bucket). Can we specify in the build command what we want to build. 
Example: 
Build SSR only. 
Build (CSR/SSG) only

3) Prioritizing Faster SSR
While static generation of (dynamic URLs) is a valid use case for some applications. Rebuilding the application with all routes takes a lot of time and is resource intensive. Generally, if a route is dynamic I would always SSR, and if a route is static (ex: sales pages, login, registration) I would always prerender. So, I don't see myself prerendering anything that is dynamic.

While caching SSR pages (ISR) and invalidating caches can improve this, personalized content (content based on the user logged in) makes caching the whole page impractical in many cases as well, not to mention the added complexity of invalidating caches. 

To me, prioritizing improving SSR performance is critical. Looking forward to streaming and improved performance of Inlining critical CSS.

**Reply by alan-agius4** [reply]
**1. Local development:** Correct, this will no longer be necessary.
**2. Build and deployment:** Generally, I wouldn't recommend this approach. In my opinion, a better workflow is to build the application once and deploy it to multiple environments accordingly. Building the application twice on two different environments seems prone to errors.
**3. Prioritizing Faster SSR:** Whilst this is important, this isn't really within the scope of this RFC.

**Comment by anfibiacreativa** [comment]
Thanks for opening for comment. I have a couple of questions,
- how would a migration from the current interim router solution to a file-based router solution work, in terms of DX. Would schematics or any other tooling be provided?
- is the type in the example meant to be a custom type or an Angular exported type? I am not sure mapping 'prerender' to SSG only, if I understand correctly, aligns with industry wide definitions and how the mechanisms work (btw, I totally understand you want to differentiate on-the-fly render from statically generated render). I would consider a different convention, if it's the second case.

**Reply by alan-agius4** [reply]
> * how would a migration from the current interim router solution to a file-based router solution work, in terms of DX. Would schematics or any other tooling be provided?

We could potentially offer a migration to transition your application to file-based routing if we choose to adopt this. NB: this configuration is not intended as a temporary solution.

> * is the type in the example meant to be a custom type or an Angular exported type? I am not sure mapping 'prerender' to SSG only, if I understand correctly, aligns with industry wide definitions and how the mechanisms work (btw, I totally understand you want to differentiate on-the-fly render from statically generated render). I would consider a different convention, if it's the second case.

The type is exported from the `@angular/ssr` package.

**Comment by yjaaidi** [comment]
**Discussion Question 3: Missing Use Cases**: How can one control the `prerender` fallback?
If a request matches a route with `prerender` mode but there is no match returned by `getPrerenderPaths()`, how can one decide to fallback to `client` mode instead of `server` mode _(considering that `server` mode is the default fallback)_?

**Reply by alan-agius4** [reply]
`getPrerenderPaths` is only required for parameterised routes that will be prendered. This is used to provide the values for the dynamic parameters during build time, not at request time.

I don't quite understand what you mean by fallback. Could you clarify?

**Reply by yjaaidi** [reply]
Here is an example. Given the following configuration:
```ts
{
  path: 'blog/:id',
  mode: 'prerender',
  getPrerenderPaths() {
    return [{id: 'a'}, {id: 'b'}];
  }
}
```

What happens if I request `/blog/c`? Will it CSR or SSR? _(that's what I meant by prerender fallback)_
How can one override that behavior?

**Reply by alan-agius4** [reply]
It depends on how you configure the routes. If `blog/:id` does not match during request time, it will fallback to the next matching route, such as `/**`. If configured as CSR, the page will be rendered on the client side (CSR), whereas if set to `server`, it will be rendered on the server side (SSR).

You could also do something like;
```ts
[
  {
    path: '/blog/:id',
    mode: 'prerender',
    getPrerenderPaths() {
      return [{ id: 'a' }, { id: 'b' }];
    },
  },
  {
    path: '/blog/**',
    mode: 'server',
  },
  {
    path: '/**',
    mode: 'prerender',
  },
];

```

**Reply by yjaaidi** [reply]
Oh ok! So the values returned by `getPrerenderPaths()` are part of the matcher. That's the part I missed.
What about query params? Will `/blog/123?q=something` match the first or second rule?

**Reply by alan-agius4** [reply]
Only the pathname is considered here. In this case, it will match the second configuration because `123` is not an ID returned by `getPrerenderPaths`. Therefore, this page will be rendered at request time.

**Reply by yjaaidi** [reply]
Oh sorry! My example was wrong. I meant `/blog/a?q=something`.
So it will match the first rule.

In this case, how can I configure this:
- `/products` should prerender
- `/products?q=...` should SSR

It is totally fine with me if this is not possible as it generally reflects that prerender is misused as a prepopulated cache where SSR should be used.

**Reply by alan-agius4** [reply]
@yjaaidi, that's not possible because the route matches based on the pathname, the query string is not taken into account in this scenario.

**Comment by yjaaidi** [comment]
**Discussion Question 4: Response / Request objects**: Are you planning to expose the whole Request API or just a subset and then extend as needed?

For example, if `Request.signal` is present, users might expect the signal to be aborted if the HTTP request is canceled.

Also, semantically speaking, the Web Request represents a request on the sending end, not the receiving end.
For instance, `Request.redirect` doesn't mean much for the receiving end so I would be surprised to have access to such a property on receiving end.

**Reply by alan-agius4** [reply]
The goal is to offer the complete Response Web Object to reduce the necessity for additional integrations aimed at standardizing and introducing another non-standard object. Numerous cloud providers already adhere to the Web Standard object, where the request is an instance of the Request class. 

In this scenario, `.signal` would be `undefined`. As for `.redirect`, it could potentially be `manual` for Node.js environments, while it would be as provided by cloud providers or other server frameworks in other environments.

**Reply by yjaaidi** [reply]
Thanks for the clarification Alan.
I assume that you meant "the complete Request Web Object" 😊

**Comment by marianocodes** [comment]
I understand the point of having separate declarations to distinguish between the server and client, but this approach also increases the learning curve and adds a potential extra file.

I believe it is more valuable to keep it simple. The mode should provide enough information to determine if it is a server or client route.

**Reply by alan-agius4** [reply]
The additional file serves a broader purpose beyond just rendering mode; it facilitates bypassing the need for relying on DCE to eliminate server-only code. This approach mitigates potential exposure of secrets when utilizing `getPrerenderPaths`.

**Reply by akaufmann** [reply]
I would also vote for only one route configuration. Couldn't we avoid the potential risks if there are modes like `server` or `prerender` - by default `client` - in the route config and the code for the required server/client route configs is rebuilt behind the scenes with the necessary risk mitigation?

**Reply by SerkanSipahi** [reply]
@alan-agius4 

That was also my first thought. Why should we have two route configurations instead of one.  I can understand that someone can conclude where which code is executed. I would prefer only one RouteConfig that can be used by both frontend and backend. We should reassess this to determine if it truly poses a security problem, and if it does, address the security issue through a different solution.

**Reply by SerkanSipahi** [reply]
We could delegate part of the task to the compiler, which removes the critical mode=server part from the route and transfers it to a server route on its own. Nothing changes for the developer. Everything happens under the hood. Furthermore, it would be an advantage if the router works the same way in the frontend as well as in the backend. So no two different ways of writing.

**Reply by alan-agius4** [reply]
Relying on Dead Code Elimination (DCE) to remove server-only code from client bundles presents a significant security risk. This is not an issue with the render mode itself, but rather with the assumption that DCE will correctly and consistently strip out sensitive server-side code. Consider the following example where a single configuration file includes server secrets:

```ts
import { Routes } from '@angular/router';

function getServerData() {
  const SERVER_SECRET = 'XXXXX';
  // Code to retrieve server data
}

export const routes: Routes = [{
  path: '',
  async getPrerenderPaths() {
    // use secret getServerData()
  },
}];
```

If DCE fails to remove the `getServerData` function, which contains the `SERVER_SECRET`, from the client bundle, it exposes sensitive information. We did discuss this with the security team at Google and relying on DCE to handle this correctly is brittle and insecure. There is a risk that server secrets could be inadvertently included in client-side code, leading to potential security vulnerabilities. Thus, it was deemed to better separate server and client code explicitly and ensure that sensitive information is never included in client-side bundles.

**Reply by marianocodes** [reply]
Question: Are there any other security concerns besides the secret? Why not consider the approach used by other frameworks, which add a prefix to expose client keys?

This might seem like a separate issue, but it also feels like the decision leans heavily on the use of secrets

**Reply by alan-agius4** [reply]
Secrets are just one part of the equation. Another crucial aspect is that `getPrerenderPaths` has the capability to directly access a database or utilize other logic and libraries that don’t necessarily need to be bundled into the client. By keeping this functionality in a separate file, we avoid depending on DCE to strip out "server-only" code.

While some frameworks rely on DCE to achieve this, its effectiveness can be inconsistent, particularly when certain coding patterns are used. Thus, when DCE fails, it can lead to two problems: an increase in the size of the client bundle and potential leakage of sensitive information if "secret" data is inadvertently included.

**Reply by SerkanSipahi** [reply]
Another possible solution would be to define the routes as usual:

```ts
export const routeConfig: RouteConfig[] = [
  { path: 'question/:id/edit', mode: 'client', },
  { path: 'login', mode: 'server', children: .... },
  { path: 'contact', mode: 'server' },
  { path: 'about-us', mode: 'server' },
];
```


In this example (see above) it should not be possible to specify `getPrerenderPaths` as long there is no critical data in the routes so that the routeConfig can be used directly in the backend (server.ts). If someone needs to expose critical data, this can only be done in the `server.ts` by enriching/merging `client.routeConfig` for specific routes with something like `getPrerenderPaths`.

The main important points are that we do not need to learn two different route configuration interfaces (client, backend) or have not to create two separate route.config files for client and server if it is unnecessary. In the described scenario, you only need to extend RouteConfig[] in server.ts when necessary, making it more of an opt-in process.

**Reply by imaksp** [reply]
Hi @SerkanSipahi , in above solution how typing could work as `getPrerenderPaths` (& other future server only fields/methods) will only be allowed in `server.ts`, so we will still need different type for server route config. & developer would still need to learn both configs.  Also in proposed solution server route config is already optional if you are not using any SSR related features. 

So IMO if DCE can't be used then this proposed solution looks better.

**Comment by Siddhu-K** [comment]
Thanks for planning this feature I wanted something like this for my app.

I was trying to setup route based rendering for my app but I couldn't get it working in angular v16. I wanted to do SSG for some static routes (for seo) and CSR for the rest of the app. I ran into issues when some of the routes which needed to be CSR only threw errors cause I was using localstorage and I couldn't figure out how to disable SSR/SSG for those routes.

- if I only want ssg+csr, Would I still need the ssr related files from `ng add @angular/ssr`

- will the issues relating to localstorage be fixed by setting the rendering mode to csr in the config?

**Reply by alan-agius4** [reply]
> if I only want ssg+csr, Would I still need the ssr related files from ng add @angular/ssr

Yes, SSG happens on the server and thus your application does need to be compatible to run on non browser environments. 

> Will the issues relating to localstorage be fixed by setting the rendering mode to csr in the config?

It depends on the usage of localstorage, if the code path is not triggered for the SSG pages than it will not error out.

**Reply by Siddhu-K** [reply]
> > if I only want ssg+csr, Would I still need the ssr related files from ng add @angular/ssr
> 
> Yes, SSG happens on the server and thus your application does need to be compatible to run on non browser environments. 
> 

I thought SSG happens during build time, how is it different from SSR if it runs on the server? (I'm not very familiar with SSR)

**Reply by alan-agius4** [reply]
By "server" in this context, I mean your build server. The application needs to run as a server application to support pre-rendering.

You can read more about this in https://angular.dev/guide/prerendering

**Comment by alfmosmq** [comment]
**Question 3**: I think there is an intermediate state between SSG and SSR. Use case (mine): a news site, publishing some articles every day. 
In such a context you want to prerender all the existing articles at build time, but you also want to prerender any new article. 
If, say, you publish 10 articles a day, currently you should rebuild 10 times a day to prerender all articles. In our current solution, we created in server.ts a custom cache, we intercept the page rendering, and save it, so next requests will find the prerendered page. We skipped prerendering at all and we built our own cache management and prerendering.
Moreover, I would add that prerendering the entire page has some issues. Consider an article page: usually, you have the article, and some box showing more articles, say "Latest news". While the article itself won't change, ever, and is the perfect candidate for prerendering, the boxes could change every time you publish a new article. 
It would be very useful if we could mark some component as "client rendered" even inside a prerendered page. Maybe with @defer used server-side to skip the rendering?

**Reply by AndrewKushnir** [reply]
> It would be very useful if we could mark some component as "client rendered" even inside a prerendered page. Maybe with `@defer` used server-side to skip the rendering?

Yes, using `@defer` for this scenario might be a good option. You can find more info on how `@defer` works on the server [here](https://angular.dev/guide/defer#behavior-with-server-side-rendering-ssr-and-static-site-generation-ssg). Another option would be to use `isPlatformBrowser` function ([docs](https://angular.dev/api/common/isPlatformBrowser)) in a component and an `@if` in a template to have different logic on the client and on the server for a location that requires client-only behavior.

**Reply by alfmosmq** [reply]
This would make the page different server-side and client-side. Wouldn't this be an issue for hydration?

**Reply by AndrewKushnir** [reply]
For the `@defer` case, Angular will hydrate the block at an initial (placeholder) state and loading of dependencies and subsequent rendering of the main template would happen on the client.

For the `@if` case with a different value on the server and on the client - Angular will be able to reconcile that by removing a server-generated part of the DOM tree (that correspond to that view) and re-creating a view from the correct `@if` branch on the client. However it's generally not recommended (ideally server and client representation should match as much as possible), so if the `@defer` approach works for your use-case, it might be a better option.

**Reply by alfmosmq** [reply]
Thanks for the clarification. Any thoughts about pre-rendering at run time?

**Reply by AndrewKushnir** [reply]
> Any thoughts about pre-rendering at run time?

If I understand your use-case correctly, it fits the Incremental Static Regeneration (ISR) scenario. ISR is outside of the scope of this proposal, but this is something that we plan to explore in the future (as mentioned in the RFC description).

**Comment by humanely** [comment]
I get error 

> ✘ [ERROR] TS2305: Module '"@angular/ssr"' has no exported member 'Request'. [plugin angular-compiler]

`
import { REQUEST, RESPONSE } from '@angular/ssr';
`

**Reply by alan-agius4** [reply]
These APIs do not exist yet. This is an RFC to about those future APIs.

**Comment by shady-ahmed-bdr** [comment]
question 3&4 :  in case of prerender  what if i have for example 2000 or even more  ids would it be too large to host on the server with all these pre rendered pages ,i think rss_feed is good for such cases.
my opinion : `{ path: 'question/:id', mode: 'prerender', }`  just serve  one SSG file and be able to send metadata belong to that route in req.body object in provide array (unlike before req, res obj were available only in SSR )

**Reply by alan-agius4** [reply]
What you're describing is somewhat like partial prerendering and is definitely interesting. However, this approach requires a server, unlike prerendering, which allows you to serve the application from a blob store.

**Reply by shady-ahmed-bdr** [reply]
in serverless hosting like cloud run SSR really takes some time to render , i was thinking to save time ,
also the user data  to  in SSG takes time for a request or localStorage to load the data and cause  layout change which some what annoying me XD unlike SSR where you can load the page with  the its data, but again its not  really optimized for such service

**Comment by jdgamble555** [comment]
This is awesome! Just a few clarifications...

1. Would `'server'` be server-only end points? It would not need to match the client in that case right?
2. Will it be possible to do some sort of server-only application like a NextJS Server Component? I'm referring to generating an application with no hydration necessary for now.
3. Would deployment bundle the routes separately? Example, a new cloud function for an endpoint etc. This is how NextJS works.

Thanks!

J

**Reply by alan-agius4** [reply]
1. No, in this case, "server" refers to a page that is rendered on the server side (SSR).
2. That is a separate request feature that is not covered by this RFC.
3. This does not affect how code is split; code splitting is managed through the Angular router route configuration.

**Reply by imaksp** [reply]
Hi, for 1. will it be rendered on the server for all cases, I mean on each client side navigation too? asking as currently it is not the case with prerender.

**Reply by jdgamble555** [reply]
So this doesn't add endpoints at all?

**Comment by Saurabh4626** [comment]
This issue is still open. Could you please provide an update on the current status?

**Comment by leonelvsc** [comment]
How event replay of ssr is going to work with secondary or multiple outlets running at the same time with mixed render types (browser and server)?, ie some dialogs with routing. Will this be supported ?, or all the events will be replayed after the ssr renders, if so browser render should use `afterRender` or `afterNextRender` right?
